File: DFA.cpp
#include "DFA.hpp"

#include <iostream>
#include <algorithm>

DFA::DFA(const NFA &nfa)
{
    std::set<unsigned int> nfaStates;
    nfaStates.insert(nfa.startState());

    std::vector<StateSet> stateSets;
    mStartState = findOrAddState(stateSets, nfa, nfaStates);

    for(unsigned int i=0; i<stateSets.size(); i++) {
        const StateSet &stateSet = stateSets[i];
        State state;

        for(const auto &pair : stateSet.transitions) {
            state.transitions[pair.first] = pair.second;
        }

        if(stateSet.nfaStates.find(nfa.acceptState()) != stateSet.nfaStates.end()) {
            mAcceptStates.insert(i);
        }

        mStates.push_back(std::move(state));
    }
}

unsigned int DFA::findOrAddState(std::vector<StateSet> &stateSets, const NFA &nfa, const std::set<unsigned int> &nfaStates)
{
    std::set<unsigned int> epsilonClosure;
    std::vector<unsigned int> queue;
    queue.insert(queue.end(), nfaStates.begin(), nfaStates.end());
    while(queue.size() > 0) {
        unsigned int state = queue.front();
        queue.erase(queue.begin());
        if(epsilonClosure.find(state) == epsilonClosure.end()) {
            epsilonClosure.insert(state);
            const auto &transitions = nfa.states()[state].epsilonTransitions;
            queue.insert(queue.end(), transitions.cbegin(), transitions.cend());
        }
    }

    for(unsigned int i=0; i<stateSets.size(); i++) {
        if(stateSets[i].nfaStates == epsilonClosure) {
            return i;
        }
    }

    std::map<Symbol, std::set<unsigned int>> transitions;
    for(unsigned int state : epsilonClosure) {
        for(const NFA::State::Transition &transition : nfa.states()[state].transitions) {
            transitions[transition.first].insert(transition.second);
        }
    }
    
    stateSets.push_back(StateSet());
    unsigned int idx = stateSets.size() - 1;
    stateSets[idx].nfaStates = epsilonClosure;
    
    for(const auto &pair : transitions) {
        stateSets[idx].transitions[pair.first] = findOrAddState(stateSets, nfa, pair.second);
    }

    return idx;
}

const std::vector<DFA::State> &DFA::states() const
{
    return mStates;
}

unsigned int DFA::startState() const
{
    return mStartState;
}

const std::set<unsigned int> &DFA::acceptStates() const
{
    return mAcceptStates;
}

void DFA::print() const
{
    std::cout << "Start state: " << mStartState << std::endl;
    std::cout << "Accept states: ";
    for(int state : mAcceptStates) {
        std::cout << state << " ";
    }
    std::cout << std::endl;
    for(int i=0; i<mStates.size(); i++) {
        std::cout << "State " << i << ":" << std::endl;
        for(const auto &pair : mStates[i].transitions) {
            std::cout << "  " << pair.first << " -> " << pair.second << std::endl;
        }
        std::cout << std::endl;
    }
}

void DFA::minimize()
{
    std::set<Symbol> alphabet;

    for(const auto &state : mStates) {
        for(const auto &transition : state.transitions) {
            alphabet.insert(transition.first);
        }
    }

    std::vector<std::set<unsigned int>> partition;

    partition.push_back(mAcceptStates);
    std::set<unsigned int> others;
    for(unsigned int i=0; i<mStates.size(); i++) {
        if(mAcceptStates.count(i) == 0) {
            others.insert(i);
        }
    }
    partition.push_back(std::move(others));

    std::vector<unsigned int> queue;
    queue.push_back(0);

    while(queue.size() > 0) {
        std::set<unsigned int> distinguisher = partition[queue.front()];
        queue.erase(queue.begin());

        for(Symbol c : alphabet) {
            std::set<unsigned int> inbound;
            for(unsigned int i=0; i<mStates.size(); i++) {
                const auto &state = mStates[i];
                const auto it = state.transitions.find(c);
                if(it != state.transitions.end() && distinguisher.count(it->second)) {
                    inbound.insert(i);
                }
            }

            if(inbound.size() == 0) {
                continue;
            }

            for(unsigned int i=0; i<partition.size(); i++) {
                std::set<unsigned int> in;
                std::set<unsigned int> out;

                for(int s : partition[i]) {
                    if(inbound.count(s)) {
                        in.insert(s);
                    } else {
                        out.insert(s);
                    }
                }

                if(in.size() > 0 && out.size() > 0) {
                    partition[i] = in;
                    partition.push_back(out);
                    unsigned int o = partition.size() - 1;

                    auto it = std::find(queue.begin(), queue.end(), i);
                    if(it == queue.end()) {
                        if(in.size() > out.size()) {
                            queue.push_back(o);
                        } else {
                            queue.push_back(i);
                        }
                    } else {
                        queue.push_back(o);
                    }
                }
            }
        }       
    }

    std::map<unsigned int, unsigned int> stateMap;
    for(unsigned int i=0; i<partition.size(); i++) {
        for(unsigned int j : partition[i]) {
            stateMap[j] = i;
        }
    }

    std::vector<State> newStates;
    for(unsigned int i=0; i<partition.size(); i++) {
        State newState;
        int s = *partition[i].begin();
        for(const auto &transition : mStates[s].transitions) {
            newState.transitions[transition.first] = stateMap[transition.second];
        }
        newStates.push_back(std::move(newState));
    }
    unsigned int newStartState = stateMap[mStartState];
    std::set<unsigned int> newAcceptStates;
    for(unsigned int s : mAcceptStates) {
        newAcceptStates.insert(stateMap[s]);
    }

    mStates = std::move(newStates);
    mStartState = newStartState;
    mAcceptStates = std::move(newAcceptStates);
}

File: src/regex/dfa.rs
use crate::util;
use util::Table;

use crate::regex;
use regex::Encoding;
use regex::NFA;

use std::collections::HashMap;
use std::collections::HashSet;
use std::collections::VecDeque;

type Symbol = usize;

#[derive(Default)]
struct State {
    pub transitions: HashMap<Symbol, usize>
}

#[derive(Default)]
struct StateSet {
    pub nfa_states: HashSet<usize>,
    pub transitions: HashMap<Symbol, usize>
}
pub struct DFA {
    num_code_points: usize,
    num_states: usize,
    pub start_state: usize,
    pub reject_state: usize,
    transitions: Table<usize>,
    accept_states: Vec<usize>
}

impl DFA {
    pub fn new(nfa: &NFA, encoding: &Encoding) -> DFA {
        let mut nfa_states = HashSet::new();
        nfa_states.insert(nfa.start_state);

        let mut state_sets = Vec::new();
        let mut states = Vec::new();
        let mut start_state = Self::find_or_add_state(&mut state_sets, nfa, &nfa_states);

        let mut accept_sets: Vec<HashSet<usize>> = vec![HashSet::<usize>::default(); nfa.accept_states.len()];
        
        for (i, state_set) in state_sets.iter().enumerate() {
            let mut state = State::default();

            for transition in &state_set.transitions {
                state.transitions.insert(*transition.0, *transition.1);
            }

            for (j, accept_state) in nfa.accept_states.iter().enumerate() {
                if state_set.nfa_states.contains(accept_state) {
                    accept_sets[j].insert(i);
                }
            }

            states.push(state);
        }

        let mut accept_states = vec![usize::MAX; states.len()];
        for (i, _) in states.iter().enumerate() {
            for (j, accept_set) in accept_sets.iter().enumerate() {
                if accept_set.contains(&i) {
                    accept_states[i] = j;
                    break;
                }
            }
        }

        (states, start_state, accept_states) = Self::minimize(&states, start_state, &accept_states);

        states.push(State::default());
        let num_states = states.len();
        let num_code_points = encoding.num_code_points();
        let reject_state = states.len() - 1;
        let mut transitions = Table::new(num_states, num_code_points, 0);

        for (i, state) in states.iter().enumerate() {
            for j in 0..num_code_points {
                if state.transitions.contains_key(&j) {
                    *transitions.at_mut(i, j) = state.transitions[&j];
                } else {
                    *transitions.at_mut(i, j) = reject_state;
                }
            }
        }

        return DFA { num_code_points, num_states, start_state, reject_state, transitions, accept_states };
    }

    pub fn transition(&self, state: usize, symbol: Symbol) -> usize {
        return *self.transitions.at(state, symbol);
    }

    pub fn accept(&self, state: usize) -> Option<usize> {
        let accept_state = self.accept_states[state];
        if accept_state == usize::MAX {
            return None;
        } else {
            return Some(accept_state);
        }
    }

    #[allow(dead_code)]
    pub fn print(&self) {
        println!("Start state: {}", self.start_state);
        println!("Accept states:");
        for (i, accept) in self.accept_states.iter().enumerate() {
            if *accept != usize::MAX {
                println!("  {i}");
            }
        }
        println!();

        for i in 0..self.num_states {
            if i == self.reject_state {
                continue;
            }

            println!("State {i}:");
            for j in 0..self.num_code_points {
                let next_state = self.transition(i, j);
                if next_state != self.reject_state {
                    println!("  {j} -> {next_state}");
                }
            }
        }
    }

    fn find_or_add_state(state_sets: &mut Vec<StateSet>, nfa: &NFA, nfa_states: &HashSet<usize>) -> usize {
        let mut epsilon_closure = HashSet::new();
        let mut queue = VecDeque::new();
        for state in nfa_states {
            queue.push_back(*state);
        }

        while queue.len() > 0 {
            let state = queue.pop_front().unwrap();
            if !epsilon_closure.contains(&state) {
                epsilon_closure.insert(state);
                for transition in &nfa.states[state].epsilon_transitions {
                    queue.push_back(*transition);
                }
            }
        }

        for (idx, state_set) in state_sets.iter().enumerate() {
            if state_set.nfa_states == epsilon_closure {
                return idx;
            }
        }

        let mut transitions: HashMap<usize, HashSet<usize>> = HashMap::new();
        for state in &epsilon_closure {
            for transition in &nfa.states[*state].transitions {
                if !transitions.contains_key(&transition.0) {
                    transitions.insert(transition.0, HashSet::<usize>::new());
                }

                transitions.get_mut(&transition.0).unwrap().insert(transition.1);
            }
        }

        state_sets.push(StateSet::default());
        let idx = state_sets.len() - 1;
        state_sets[idx].nfa_states = epsilon_closure;

        for transition in &transitions {
            let s = Self::find_or_add_state(state_sets, nfa, transition.1);
            state_sets.get_mut(idx).unwrap().transitions.insert(*transition.0, s);
        }

        return idx;
    }

    fn minimize(states: &[State], start_state: usize, accept_states: &[usize]) -> (Vec<State>, usize, Vec<usize>) {
        let mut alphabet = HashSet::new();
        for state in states.iter() {
            for transition in &state.transitions {
                alphabet.insert(*transition.0);
            }
        }

        let mut accept_sets: HashMap<usize, HashSet<usize>> = HashMap::new();
        for (i, accept_state) in accept_states.iter().enumerate() {
            if !accept_sets.contains_key(accept_state) {
                accept_sets.insert(*accept_state, HashSet::new());
            }
            accept_sets.get_mut(accept_state).unwrap().insert(i);
        }

        let mut partition = Vec::new();
        for set in accept_sets.into_values() {
            partition.push(set);
        }
        
        let mut queue = VecDeque::new();
        for i in 0..partition.len() {
            queue.push_back(i);
        }

        while queue.len() > 0 {
            let s = queue.pop_front().unwrap();
            let distinguisher = partition[s].clone();
                    
            for c in &alphabet {
                let mut inbound = HashSet::new();
                for (i, state) in states.iter().enumerate() {
                    if state.transitions.contains_key(c) && distinguisher.contains(&state.transitions[c]) {
                        inbound.insert(i);
                    }
                }

                if inbound.is_empty() {
                    continue;
                }

                for i in 0..partition.len() {
                    let mut in_set = HashSet::new();
                    let mut out_set = HashSet::new();

                    for s in &partition[i] {
                        if inbound.contains(s) {
                            in_set.insert(*s);
                        } else {
                            out_set.insert(*s);
                        }
                    }

                    if in_set.len() > 0 && out_set.len() > 0 {
                        partition[i] = in_set.clone();
                        partition.push(out_set.clone());
                        let o = partition.len() - 1;

                        if queue.contains(&i) {
                            queue.push_back(o);
                        } else {
                            if in_set.len() > out_set.len() {
                                queue.push_back(o);
                            } else {
                                queue.push_back(i);
                            }
                        }
                    }
                }
            }
        }

        let mut state_map = HashMap::new();
        for (i, part) in partition.iter().enumerate() {
            for j in part {
                state_map.insert(*j, i);
            }
        }

        let mut new_states = Vec::new();
        for part in &partition {
            let mut new_state = State::default();
            let s = part.iter().next().unwrap();
            for transition in &states[*s].transitions {
                new_state.transitions.insert(*transition.0, state_map[transition.1]);
            }
            new_states.push(new_state);
        }

        let new_start_state = state_map[&start_state];
        let mut new_accept_states = vec![usize::MAX; new_states.len()];
        for i in 0..states.len() {
            new_accept_states[state_map[&i]] = accept_states[i];
        }

        return (new_states, new_start_state, new_accept_states);
    }
}